Use ServiceCaller to access JAASService

tjwatson opened the pull request:
This avoids overflowing the usecount of the OSGi service and simplifies getting the service over and over.

tjwatson committed:
Use ServiceCaller to access JAASService

This avoids overflowing the usecount of the OSGi service
and simplifies getting the service over and over.

tjwatson commented:
#build

dazavala commented on dev/com.ibm.ws.security.authentication.builtin/src/com/ibm/ws/security/authentication/jaas/modules/LoginModuleHelper.java:
```diff
@@ -1,28 +1,37 @@
+        ServiceCaller<JAASService> temp;
+        try {
+            temp = new ServiceCaller<>(LoginModuleHelper.class, JAASService.class);
+        } catch (IllegalStateException e) {
```
Looks good.  ServiceCaller(<init>) should never throw NPE as neither parameter can be null.

dazavala commented on dev/com.ibm.ws.security.authentication.builtin/src/com/ibm/ws/security/authentication/jaas/modules/LoginModuleHelper.java:
```diff
@@ -34,8 +43,7 @@ public static JAASService getJAASService() {
 
-        BundleContext bc = FrameworkUtil.getBundle(IdentityAssertionLoginModule.class).getBundleContext();
-        return bc.getService(bc.getServiceReference(JAASService.class));
+        return jaasService.current().orElseThrow(() -> new NullPointerException("No JAASService found."));
```
It is possible the `jaasService` field may be null here, causing a plain ol' NPE when calling `.current()`.  The field may be null only when `ServiceCaller(<init>)` throws an `ISE` due to not obtaining the bundle for class `LoginModuleHelper`.  This is extremely unlikely and would otherwise indicate more severe problems. No change requested.

tjwatson replied:
Right, I may (later) change the ServiceCaller to not throw an ISA on construction here.  Instead, I would throw ISA when tracking the service the first time the BundleContext is needed from the Bundle.

tjwatson replied:
I changed the constructor to no longer throw an ISA.  This makes the unit tests scenario more simple and also allows the code here to be simplified for constructing the ServiceCaller.

dazavala replied:
+1 on improved usability.

dazavala reviewed (dismissed):
@tjwatson Looks good.  I posted a couple comments about your solution. No changes requested.

una-tapa reviewed (dismissed):
Looks good to me.

una-tapa commented:
fyi @utle in case you have a cycle to review.

tjwatson committed:
Fix ServiceCaller to use highest rank service

Also avoid IllegalStateException on construction.
Instead throw ISE on first use when the BundleContext
cannot be determined.

tjwatson commented:
I have updated the PR to fix some issues with `ServiceCaller`

1. Needed to pull in the fix I did for Equinox for the service ranking
2. Changed the `ServiceCaller` constructor to no longer throw an ISE.  Instead ISE is thrown on first use.  This simplifies constructing the `ServiceCaller` in a constant of a class and allowing the class to be exercised outside of OSGi

dazavala reviewed (approved):
LGTM.  Didn't see potential for NPE due to null `bundle` in class `ReferenceAndService` as  instances of `ReferenceAndService` are constructed after the call to `getContext()`. Also, all public methods that may throw the ISE are documented: `call(Consumer<S>)`, `call(Function<S, R>)`, and `current()`.

una-tapa reviewed (approved):
Thanks for the update.

una-tapa commented:
@utle - I created 
https://github.com/OpenLiberty/open-liberty/issues/28270
to continue the discussion on the message.

tjwatson commented:
#libby
